feat: 로또 구매 구현

songkg7 opened the pull request:
## ✏️ 구현 내용

- [x] 구매할 로또 개수 판단을 위한 `LottoCount` 객체를 생성합니다
- [x] `Money` 객체에 operator function 을 정의하여 사칙연산이 수행가능하도록 수정합니다
- [ ] ~중복되지 않은 6자리 번호를 입력하여 `LottoTicket` 객체를 생성합니다~
- [x] `LottoNumbers` 일급 컬렉션 객체를 구성하여 6자리의 로또 번호 묶음을 구현합니다
- [x] 예외 메세지들의 scope 를 일부 조절합니다
- [x] 예외 발생시 루프를 위한 top level function 의 예외를 조절합니다
	- 다양한 예외 케이스를 캐치하기 위함입니다
- [x] codecov 에서 view 패키지는 표시하지 않도록 하는 설정파일을 추가합니다
- [x] ktlint 가 좀 더 명확하게 표시되도록 일부 옵션을 editorconfig 에 직접 정의합니다

## 🤔 논의하고 싶은 내용 / 질문 등

- ktlint 는 **가독성보다는 일관성**을 우선시하고 있어서 그런지 생각보다 꽤 깐깐하네요
- 생각보다 로또 수동 구매 부분에서 고민해야할 부분이 많아서 어느 정도 가이드만 구현하는걸로 마무리했습니다
- 로또 번호 입력 검증 등의 예외처리가 꼼꼼하게 안되어 있고 다소 러프하게 구현되어 있는데 깔끔하게 정리해주셨으면 좋겠네요~

songkg7 committed:
ci: add ignore file for codecov

songkg7 committed:
refactor: message 정리

songkg7 committed:
style: unnecessary trailing comma in test

songkg7 committed:
feat: 구매할 로또의 개수를 결정하는 LottoCount 구현

songkg7 committed:
fix: input 검증을 위한 try catch loop 예외 레벨 조정

songkg7 committed:
style: trailing comma 정의

songkg7 committed:
style: multiline expression wrapping

songkg7 committed:
style: new line function signature

songkg7 committed:
impv: plus, minus 연산 추가

doxxx93 commented:
아직 진행중이시지만, 
- 기존에 제가 작성했던 companion object의 사용보다 그냥 상수화 시키는 것이 더 깔끔해보이네요 ㅎ_ㅎ
- [operator overloading](https://kotlinlang.org/docs/operator-overloading.html)에 대한 사용도 정말 적합해보여요 좋은 가독성도 덤이고, 덕분에 테스트 코드에서 `*` 연산을 통해 직관적으로 작성할 수 있군요
- 그 외의 린트 관련된 설정을 잘 봐주셔서 일관성있는 코드 작성에 도움이 되겠어요

songkg7 commented:
> 아직 진행중이시지만,
> 
> * 기존에 제가 작성했던 companion object의 사용보다 그냥 상수화 시키는 것이 더 깔끔해보이네요 ㅎ_ㅎ
> * [operator overloading](https://kotlinlang.org/docs/operator-overloading.html)에 대한 사용도 정말 적합해보여요 좋은 가독성도 덤이고, 덕분에 테스트 코드에서 `*` 연산을 통해 직관적으로 작성할 수 있군요
> * 그 외의 린트 관련된 설정을 잘 봐주셔서 일관성있는 코드 작성에 도움이 되겠어요

벌써 확인하셨군요 ㅎㅎ 나중에 언급해도 될 부분이긴 하지만 사실 operator 를 overloading 하는거는 굉장히 주의해야 하는 부분이에요. 내부 구현을 감추기 때문에 메서드가 호출된건지 그냥 연산자가 호출된건지를 혼란스럽게 할 수 있기 때문이에요. 그래서 operator 를 사용하기보다는 그냥 `money.div(other)` 처럼 메서드 호출 방식으로 쓰는걸 권장하는 곳도 있습니다.

이번에는 객체가 Money 라는 숫자를 쉽게 연상할 수 있는 객체였기 때문에 혼란이 적을 것이라 생각해서 사칙연산을 구현해봤습니다.

songkg7 committed:
Add LottoNumbers model and update View and Controller

A feature to validate the number of lotto numbers was created. For this, a LottoNumbers model was added with a from method to transform a list of Integers into LottoNumbers and validate the size. The View is updated to request and display useful lotto-related information. The Controller is also updated to include interactions with the LottoNumbers model. Now, the application is able to manage the user's lotto numbers directly and enforce the correct number of lotto entries.

songkg7 committed:
style: code formatting

songkg7 committed:
ci: ignore LottoApp

doxxx93 commented:
input view와.. output view, 그리고 예외 상황이 발생했을 때 메시지를 어디에 둘지도 꽤 고민이 되는 거 같습니다.. ㅋㅋ

songkg7 commented:
> input view와.. output view, 그리고 예외 상황이 발생했을 때 메시지를 어디에 둘지도 꽤 고민이 되는 거 같습니다.. ㅋㅋ

그러게요. 사실 굳이 모아두지 않고 사용하는 곳에 메세지를 두는 편이긴한데 뭔가 view 라고 있으니 메세지도 모여있어야할거 같고~

this-is-spear commented on src/test/kotlin/model/lotto/LottoCountTest.kt:
```diff
@@ -0,0 +1,51 @@
+
+class LottoCountTest : FreeSpec(
+    {
+        "수동 로또 구매 개수를 입력할 때" - {
```
수동 로또 구매 개수를 입력할 때 음수를 입력하면 문제가 발생하지 않을까 우려됩니다!

> 경계 값 분석만큼 테스트 비용을 높이는 친구는 없다고 생각이 들어요 ㅎㅎ 그래서 일부 자료구조는 일정한 프로세스를 가지고 접근하고 있어요 😄
>예를 들어, 저는 정수를 테스트 할 때 `양수`, `0`, `음수`, `조건` 을 테스트하고 있습니다.

this-is-spear replied:
> 로또 번호 입력 검증 등의 예외처리가 꼼꼼하게 안되어 있고 다소 러프하게 구현되어 있는데 깔끔하게 정리해주셨으면 좋겠네요~

헉 제가 PR 코멘트를 읽지 않았군요 ㅎㅎ 머지된다면 제가 검증 코드를 추가하겠습니다 ☺️

doxxx93 replied:
저도 PR 코멘트와 테스트 코드에서 not implemented를 보고 남겨두신 부분이구나 했습니다!

this-is-spear commented on src/main/kotlin/model/lotto/LottoNumbers.kt:
```diff
@@ -0,0 +1,13 @@
+package model.lotto
+
+import view.INVALID_LOTTO_NUMBER_COUNT_MESSAGE
```
해당 메시지는 `LottoNumbers` 에서 관리하는건 어떨까요 ㅎㅎ
`view` 레이어와 `model` 레이어의 의존 방향이 단방향이면 관리가 편해보여요!

doxxx93 replied:
이 부분에 대해서 저도 코멘트로 남겨봤었는데, 정확하게는 어떻게 처리하는 게 좋을지 애매한 것 같았습니다. ㅎ_ㅎ

건창님께서 한번 도전해보심이 어떨까요?

this-is-spear replied:
좋습니다! 🙌

this-is-spear reviewed (approved):
수고하셨습니다! 이번 기회에 `operator keyword`를 맛 볼 수 있었습니다 ㅎㅎ

doxxx93 commented:
@this-is-spear 저도 approve했습니다. 남은 작업들을 건창님이 멋지게 구현해보시는 걸 기대하겠습니다!
